src/Microsoft.AspNetCore.Razor.Language/Legacy/WhiteSpaceRewriter.cs

```diff
@@ -1,41 +1,65 @@
 
-using System.Collections.Generic;
 using System.Linq;
+using Microsoft.AspNetCore.Razor.Language.Syntax;
 
 namespace Microsoft.AspNetCore.Razor.Language.Legacy
 {
-    internal class WhiteSpaceRewriter : MarkupRewriter
+    internal class WhitespaceRewriter : SyntaxRewriter
     {
-        protected override bool CanRewrite(Block block)
+        public override SyntaxNode Visit(SyntaxNode node)
         {
-            return block.Type == BlockKindInternal.Expression && Parent != null;
+            var result = node;
+            var children = node.ChildNodes();
+            for (var i = 0; i < children.Count; i++)
+            {
+                var child = children[i];
+                if (child is CSharpCodeBlockSyntax codeBlock &&
+                    TryRewriteWhitespace(codeBlock, out var rewritten, out var whitespaceLiteral))
+                {
+                    // Replace the existing code block with the whitespace literal
+                    // followed by the rewritten code block (with the code whitespace removed).
+                    result = result.ReplaceNode(codeBlock, new SyntaxNode[] { whitespaceLiteral, rewritten });
+                }
+            }
+
+            return base.Visit(result);
         }
 
-        protected override SyntaxTreeNode RewriteBlock(BlockBuilder parent, Block block)
+        private bool TryRewriteWhitespace(CSharpCodeBlockSyntax codeBlock, out CSharpCodeBlockSyntax rewritten, out SyntaxNode whitespaceLiteral)
         {
-            var newBlock = new BlockBuilder(block);
-            newBlock.Children.Clear();
-            var ws = block.Children.FirstOrDefault() as Span;
-            IEnumerable<SyntaxTreeNode> newNodes = block.Children;
-            if (ws.Content.All(char.IsWhiteSpace))
+            // Rewrite any whitespace represented as code at the start of a line preceding an expression block.
+            // We want it to be rendered as Markup.
+
+            rewritten = null;
+            whitespaceLiteral = null;
+            var children = codeBlock.ChildNodes();
+            if (children.Count < 2)
             {
-                // Add this node to the parent
-                var builder = new SpanBuilder(ws);
-                builder.ClearTokens();
-                FillSpan(builder, ws.Start, ws.Content);
-                parent.Children.Add(builder.Build());
-
-                // Remove the old whitespace node
-                newNodes = block.Children.Skip(1);
+                return false;
             }
 
-            foreach (SyntaxTreeNode node in newNodes)
+            if (children[0] is CSharpStatementLiteralSyntax literal &&
+                (children[1] is CSharpExpression || children[1] is CSharpImplicitExpression))
             {
-                newBlock.Children.Add(node);
+                var literalContent = GetContent(literal);
+                if (literalContent.All(char.IsWhiteSpace))
```

Comment:
Should this be `string.IsWhitespace(`? This is a really inefficient way to implement that kind of check.

Reply:
This line was copy-paste from old code. Will replace.